OFEP-single-context-paradigm.md

```diff
@@ -0,0 +1,105 @@
+## OFEP-single-context-paradigm
+
+## State: DRAFTING
+
+This draft outlines a "single-context" paradigm - an alternative pattern and set of APIs supporting client use-cases.
+
+## Background
+
+In contrast with server-side or other service-type applications, client side apps typically operate in the context of a single user.
+Most feature flagging SDKs for these applications have been designed with this in mind.
+In summary, most client/web SDKs operate something like this:
+
+- an initialization occurs, which fetches evaluated flags in bulk for a given context (user)
+- the evaluated flags are cached in the SDK
+- flag evaluations take place against this cache, without a need to provide context (context was already used to evaluate flags in bulk)
+- Functions/methods are exposed on the SDK that signal the cache is no longer valid, and must be reconciled based on a context change. This frequently involves a network request or I/O operation.
+
+This paradigm doesn't fit well with the existing SDK, which, like most server-side SDKs, emphasizes realtime evaluation and context-per-evaluation.
+
+Though not all client side SDKs function in this way, those that do allow context per evaluation can conform to this model fairly easily.
+
+## Design
+
+To better support the "single-context" paradigm we need to define some additional handlers and add some flexibility to current APIs.
```

Comment:
How do these API changes exist alongside the current APIs? Does an author have to choose which to follow or is it possible to incorporate both designs?

Reply:
I think it would be possible (though perhaps not recommended) to follow both. I think in most cases, client-focused SDKs would implement the static-context paradigm, while server/multi-user applications would do dynamic context. We could add guidance in the docs about this.